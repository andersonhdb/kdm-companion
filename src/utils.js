import R from 'ramda'
import { values } from 'mobx'

export const capitalize = str => str[0].toUpperCase() + str.slice(1)

export function uuid() {
  return ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(
    /[018]/g,
    a => (a ^ ((Math.random() * 16) >> (a / 4))).toString(16)[0]
  )
}

export function expansionFilter(map, expansion) {
  return R.filter(item => item.expansion.id === expansion.id, values(map))
}

/**
 *
 * @param {object} base object to merge data into (eg: getSnapshot(SettlementStat.create()))
 * @param {array} values array of objects with base attributes to merge (eg: values(self.innovations))
 * @param {array} attribute attribute from each values item to pick to merge into the base (eg: ['settlement', 'newborn'] )
 */
export function objectMerge(base, values, attribute) {
  let keys = Object.keys(base)
  let result = values.reduce((merged, item) => {
    if (R.path(attribute, item)) {
      // combine the 2 objects: current accumulated base and the one from the values item.
      return keys.reduce((keysAcc, key) => {
        return {
          ...keysAcc,
          [key]:
            typeof merged[key] === 'string'
              ? R.path(attribute, item)[key].length > 0
                ? merged[key] + R.path(attribute, item)[key] + '\n'
                : merged[key]
              : merged[key] + R.path(attribute, item)[key],
        }
      }, {})
    } else {
      return merged
    }
  }, base)

  return result
}
