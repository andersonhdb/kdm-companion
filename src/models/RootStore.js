import { types, getSnapshot } from 'mobx-state-tree'
import { SettlementLocation } from './SettlementLocation'
import { Campaign } from './Campaign'
import { Innovation } from './Innovation'
import { Resource } from './Resource'
import { Monster } from './Monster'
import { Gear } from './Gear'
import { Expansion } from './Expansion'
import { Principle } from './Principle'
import { FightingArt, SecretFightingArt } from './FightingArt'
import { Disorder } from './Disorder'
import { Ability } from './Ability'
import { WeaponProficiency } from './WeaponProficiency'
import { Subscription } from './Subscription'
import { uuid } from '../utils'
import R from 'ramda'

import locationsData from '../data/settlement_locations'
import innovationsData from '../data/innovations'
import fightingArtsData from '../data/fightingArts'
import secretFightingArtsData from '../data/secretFightingArts'
import disordersData from '../data/disorders'
import resourceData from '../data/resources'
import monsterData from '../data/monsters'
import gearData from '../data/gear'
import expansionData from '../data/expansions'
import principlesData from '../data/principles'
import abilitiesData from '../data/abilities'
import weaponProficiencyData from '../data/weaponProficiencies'

export default types
  .model('RootStore', {
    locations: types.optional(types.map(SettlementLocation), locationsData),
    innovations: types.optional(types.map(Innovation), innovationsData),
    fightingArts: types.optional(types.map(FightingArt), fightingArtsData),
    secretFightingArts: types.optional(
      types.map(SecretFightingArt),
      secretFightingArtsData
    ),
    disorders: types.optional(types.map(Disorder), disordersData),
    resources: types.optional(types.map(Resource), resourceData),
    monsters: types.optional(types.map(Monster), monsterData),
    gear: types.optional(types.map(Gear), gearData),
    expansions: types.optional(types.map(Expansion), expansionData),
    principles: types.optional(types.map(Principle), principlesData),
    abilities: types.optional(types.map(Ability), abilitiesData),
    weaponProficiencies: types.optional(
      types.map(WeaponProficiency),
      weaponProficiencyData
    ),
    // stuff stored to disc after this
    version: 1,
    campaigns: types.optional(types.array(Campaign), [
      {
        id: 'new',
      },
    ]),
    selectedCampaign: types.optional(types.reference(Campaign), 'new'),
    subscription: types.optional(Subscription, {}),
  })
  .actions(self => ({
    createCampaign(name) {
      id = uuid()
      self.campaigns.push({ id, settlement: { name: name, survivalLimit: 1 } })
      self.selectCampaign(id)
    },
    selectCampaign(id) {
      self.selectedCampaign = id
    },
    deleteCampaign(id) {
      self.campaigns = R.reject(c => c.id === id, self.campaigns)
      if (self.campaigns.length === 0) {
        self.createCampaign('New Settlement')
      } else {
        self.selectedCampaign = self.campaigns[0]
      }
    },
    setVersion(version) {
      self.version = version
    },
    load(data) {
      self.campaigns = data.campaigns
      self.selectedCampaign = data.selectedCampaign
      self.version = data.version || 1
      self.subscription = data.subscription || {}
    },
  }))
  .views(self => ({
    get data() {
      return {
        campaigns: getSnapshot(self.campaigns),
        selectedCampaign: self.selectedCampaign.id,
        version: self.version,
        subscription: getSnapshot(self.subscription),
      }
    },
    get campaignExpansions() {
      return self.selectedCampaign.expansionList
    },
    selectedExpansionFilter(map) {
      return self.selectedCampaign.selectedExpansionFilter(map)
    },
    get availableLocations() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.locations)
      )
    },
    get availableInnovations() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.innovations)
      )
    },
    get availableFightingArts() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.fightingArts)
      )
    },
    get availableSecretFightingArts() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.secretFightingArts)
      )
    },
    get availableDisorders() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.disorders)
      )
    },
    get availableAbilities() {
      return R.sortBy(
        R.prop('name'),
        self.selectedExpansionFilter(self.abilities)
      )
    },
    get availableHunts() {
      return self.selectedExpansionFilter(self.monsters)
    },
    get availableWeaponProficiencies() {
      return self.selectedExpansionFilter(self.weaponProficiencies)
    },
    get numCampaigns() {
      return self.campaigns.length
    },
  }))
