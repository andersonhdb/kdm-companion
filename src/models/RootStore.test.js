import { getSnapshot } from 'mobx-state-tree'
import RootStore from './RootStore'
import * as utils from '../utils'

afterEach(() => {
  jest.restoreAllMocks()
})

it('initializes', () => {
  const store = RootStore.create()
})

describe('.load', () => {
  let store

  beforeEach(() => {
    store = RootStore.create()
  })

  it('handles empty data', () => {
    store.load({})
    expect(store.data).toMatchSnapshot()
  })

  it('loads saved data', () => {
    let data = {
      campaigns: [
        {
          bonuses: {},
          endeavors: {},
          expansions: {
            core: 'core',
          },
          hunting: null,
          id: 'new',
          innovations: {},
          locations: {},
          notes: 'my notes',
          principles: {
            conviction: null,
            death: null,
            newlife: null,
            society: null,
          },
          settlement: {
            departing: null,
            name: 'New Settlement',
            newborn: null,
            showdown: null,
            survivalLimit: 3,
            survivors: {},
          },
          showdown: null,
          stored_resources: {},
        },
      ],
      selectedCampaign: 'new',
      subscription: {
        appLastActiveAt: 0,
        autoRenewing: false,
        checkedAt: 0,
        purchasedAt: 0,
        response: '',
        transactionId: undefined,
      },
      version: 1,
    }
    store.load(data)
    expect(store.data).toMatchSnapshot()
  })
})

it('all in', () => {
  const store = RootStore.create()

  let uuidMock = jest.spyOn(utils, 'uuid')
  uuidMock.mockImplementation(() => '960aa7ae-598a-49eb-888b-c6baa0006751')
  store.createCampaign('test')

  // add all expansions
  //TODO: read data file and activate all, same for the other stuff.
  store.selectedCampaign.selectExpansion({ id: 'dbk' })
  store.selectedCampaign.selectExpansion({ id: 'dk' })
  store.selectedCampaign.selectExpansion({ id: 'fk' })
  store.selectedCampaign.selectExpansion({ id: 'gka' })
  store.selectedCampaign.selectExpansion({ id: 'gorm' })
  store.selectedCampaign.selectExpansion({ id: 'lg' })
  store.selectedCampaign.selectExpansion({ id: 'lk' })
  store.selectedCampaign.selectExpansion({ id: 'lt' })
  store.selectedCampaign.selectExpansion({ id: 'manhunter' })
  store.selectedCampaign.selectExpansion({ id: 'slenderman' })
  store.selectedCampaign.selectExpansion({ id: 'spidicules' })
  store.selectedCampaign.selectExpansion({ id: 'sunstalker' })

  store.selectedCampaign.selectLocation({ id: 'mask_maker' })
  store.selectedCampaign.selectLocation({ id: 'the sun' })

  expect(store.data).toMatchSnapshot()
})
