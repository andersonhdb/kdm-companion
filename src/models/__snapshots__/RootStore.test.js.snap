// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.load handles empty data 1`] = `
Object {
  "campaigns": Array [
    Object {
      "bonuses": Object {},
      "endeavors": Object {},
      "expansions": Object {
        "core": "core",
      },
      "hunting": null,
      "id": "new",
      "innovations": Object {},
      "locations": Object {},
      "notes": "",
      "principles": Object {
        "conviction": null,
        "death": null,
        "newlife": null,
        "society": null,
      },
      "settlement": Object {
        "departing": null,
        "name": "New Settlement",
        "newborn": null,
        "showdown": null,
        "survivalLimit": 1,
        "survivors": Object {},
      },
      "showdown": null,
      "stored_resources": Object {},
    },
  ],
  "selectedCampaign": "new",
  "subscription": Object {
    "appLastActiveAt": 0,
    "autoRenewing": false,
    "checkedAt": 0,
    "purchasedAt": 0,
    "response": "",
    "transactionId": undefined,
  },
  "version": 1,
}
`;

exports[`.load loads saved data 1`] = `
Object {
  "campaigns": Array [
    Object {
      "bonuses": Object {},
      "endeavors": Object {},
      "expansions": Object {
        "core": "core",
      },
      "hunting": null,
      "id": "new",
      "innovations": Object {},
      "locations": Object {},
      "notes": "my notes",
      "principles": Object {
        "conviction": null,
        "death": null,
        "newlife": null,
        "society": null,
      },
      "settlement": Object {
        "departing": null,
        "name": "New Settlement",
        "newborn": null,
        "showdown": null,
        "survivalLimit": 3,
        "survivors": Object {},
      },
      "showdown": null,
      "stored_resources": Object {},
    },
  ],
  "selectedCampaign": "new",
  "subscription": Object {
    "appLastActiveAt": 0,
    "autoRenewing": false,
    "checkedAt": 0,
    "purchasedAt": 0,
    "response": "",
    "transactionId": undefined,
  },
  "version": 1,
}
`;

exports[`all in 1`] = `
Object {
  "campaigns": Array [
    Object {
      "bonuses": Object {},
      "endeavors": Object {},
      "expansions": Object {
        "core": "core",
      },
      "hunting": null,
      "id": "new",
      "innovations": Object {},
      "locations": Object {},
      "notes": "",
      "principles": Object {
        "conviction": null,
        "death": null,
        "newlife": null,
        "society": null,
      },
      "settlement": Object {
        "departing": null,
        "name": "New Settlement",
        "newborn": null,
        "showdown": null,
        "survivalLimit": 1,
        "survivors": Object {},
      },
      "showdown": null,
      "stored_resources": Object {},
    },
    Object {
      "bonuses": Object {},
      "endeavors": Object {
        "build_bone_smith": Object {
          "expansion": "sunstalker",
          "id": "build_bone_smith",
          "name": "Build - Bone Smith",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
            ],
            "location": "the sun",
            "not_innovation": null,
            "not_location": "bone_smith",
          },
        },
        "build_organ_grinder": Object {
          "expansion": "sunstalker",
          "id": "build_organ_grinder",
          "name": "Build - Organ Grinder",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
            ],
            "location": "the sun",
            "not_innovation": null,
            "not_location": "organ_grinder",
          },
        },
        "build_skinnery": Object {
          "expansion": "sunstalker",
          "id": "build_skinnery",
          "name": "Build - Skinnery",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
            ],
            "location": "the sun",
            "not_innovation": null,
            "not_location": "skinnery",
          },
        },
        "golden_eyed_king": Object {
          "expansion": "core",
          "id": "golden_eyed_king",
          "name": "Hunt the Golden Eyed King",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
              Object {
                "gear": "phoenix_mask",
                "keyword": null,
                "quantity": 1,
                "resource": null,
              },
            ],
            "location": "mask_maker",
            "not_innovation": null,
            "not_location": null,
          },
        },
        "great_golden_cat": Object {
          "expansion": "core",
          "id": "great_golden_cat",
          "name": "Hunt the Great Golden Cat",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
              Object {
                "gear": "white_lion_mask",
                "keyword": null,
                "quantity": 1,
                "resource": null,
              },
            ],
            "location": "mask_maker",
            "not_innovation": null,
            "not_location": null,
          },
        },
        "innovate": Object {
          "expansion": "sunstalker",
          "id": "innovate",
          "name": "Innovate",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
              Object {
                "gear": null,
                "keyword": "organ",
                "quantity": 1,
                "resource": null,
              },
              Object {
                "gear": null,
                "keyword": "bone",
                "quantity": 1,
                "resource": null,
              },
              Object {
                "gear": null,
                "keyword": "hide",
                "quantity": 1,
                "resource": null,
              },
            ],
            "location": "the sun",
            "not_innovation": null,
            "not_location": null,
          },
        },
        "mad_steed": Object {
          "expansion": "core",
          "id": "mad_steed",
          "name": "Hunt the Mad Streed",
          "recipe": Object {
            "innovation": null,
            "items": Array [
              Object {
                "gear": null,
                "keyword": null,
                "quantity": 1,
                "resource": "endeavor",
              },
              Object {
                "gear": "antelope_mask",
                "keyword": null,
                "quantity": 1,
                "resource": null,
              },
            ],
            "location": "mask_maker",
            "not_innovation": null,
            "not_location": null,
          },
        },
      },
      "expansions": Object {
        "core": "core",
        "dbk": "dbk",
        "dk": "dk",
        "fk": "fk",
        "gka": "gka",
        "gorm": "gorm",
        "lg": "lg",
        "lk": "lk",
        "lt": "lt",
        "manhunter": "manhunter",
        "slenderman": "slenderman",
        "spidicules": "spidicules",
        "sunstalker": "sunstalker",
      },
      "hunting": null,
      "id": "960aa7ae-598a-49eb-888b-c6baa0006751",
      "innovations": Object {},
      "locations": Object {
        "mask_maker": "mask_maker",
        "the sun": "the sun",
      },
      "notes": "",
      "principles": Object {
        "conviction": null,
        "death": null,
        "newlife": null,
        "society": null,
      },
      "settlement": Object {
        "departing": null,
        "name": "test",
        "newborn": null,
        "showdown": null,
        "survivalLimit": 1,
        "survivors": Object {},
      },
      "showdown": null,
      "stored_resources": Object {},
    },
  ],
  "selectedCampaign": "960aa7ae-598a-49eb-888b-c6baa0006751",
  "subscription": Object {
    "appLastActiveAt": 0,
    "autoRenewing": false,
    "checkedAt": 0,
    "purchasedAt": 0,
    "response": "",
    "transactionId": undefined,
  },
  "version": 1,
}
`;
